### PR TITLE
Use GemmaAttention for Gemma

### PR DESCRIPTION
**Pull request by qihqi**
This way it produces more accurate results (with EOS)

{'rouge1': 36.9881, 'rouge2': 13.3464, 'rougeL': 21.7437, 'rougeLsum': 35.1489, 'gen_len': 1295948, 'gen_num': 1000}

### CODE REVIEW

File: convert_checkpoints.py [hunk 1]
```diff
@@ -447,10 +448,10 @@ def convert_hf_gemma_weights(
       state_dict[new_key.replace("qkv_proj", "wk")] = k
       state_dict[new_key.replace("qkv_proj", "wv")] = v
       continue
-    if "o_proj" in key:
-      new_key = new_key.replace("o_proj", "wo")
+
     if new_key != key:
       state_dict[new_key] = state_dict.pop(key)
+  output_ckpt_dir.mkdir(parents=True, exist_ok=True)
```
Comment (lsy323) [inline convert_checkpoints.py:454]: This shouldn't be needed. Output folder will be created in `_export_to_local` https://github.com/google/jetstream-pytorch/blob/811d718c1f93e5ce37182e2c1ec54d3dc0b4aed7/convert_checkpoints.py#L355

File: convert_checkpoints.py [hunk 2]
```diff
@@ -414,11 +414,12 @@ def convert_hf_gemma_weights(
   ckpt_file = list(input_ckpt_dir.glob("*.ckpt"))
   assert len(ckpt_file) == 1, "only expect 1 ckpt file for Gemma model."
   ckpt_file = ckpt_file[0]
-  state_dict = torch.load(ckpt_file, map_location=torch.device("cpu"))[
+  state_dict = torch.load(str(ckpt_file), map_location=torch.device("cpu"))[
       "model_state_dict"
   ]
   model_config = json.loads((input_ckpt_dir / "config.json").read_text())
   for key in list(state_dict.keys()):
+    print(key)
```
Comment (lsy323) [inline convert_checkpoints.py:422]: remove?

**Review by lsy323 — APPROVED** [review]

File: default_shardings/gemma.yaml [hunk 3]
```diff
@@ -3,7 +3,9 @@
 # "replicated" to signify "replicated".
 # Integer signify axis to shard: 0 <= shard axis < rank
 
-freqs_cis : -1 #  torch.complex64 (16384, 128)
+freqs_cis : null #  torch.complex64 (16384, 128)
+layers.*.self_attn.qkv_proj.weight: 0
```
Comment (FanhaiLu1) [inline default_shardings/gemma.yaml:7]: Is this one only for test purpose? In gemma model, I saw the code directly read wq,wk and wv.

File: jetstream_pt/layers.py [hunk 4]
```diff
@@ -219,81 +336,8 @@ def forward(
...
-        )
-        if seqlen == 1:
-          output = output[:, :, 0:1, :]
-        # output = torch.matmul(scores, values)
-        self.env.apply_sharding(output, axis=1)
-        output = output.transpose(-3, -2).contiguous().view(bsz, seqlen, -1)
-        self.env.apply_sharding(output, axis=2)
-      return self.wo(output)
+    output = self.attention_kernel(xq, xk, xv, mask, cache)
+    output = output.transpose(-3, -2).contiguous().view(bsz, seqlen, -1)
```
Comment (FanhaiLu1) [inline jetstream_pt/layers.py:342]: Nice, code is cleaner with refactoring the attention kernel.

File: jetstream_pt/third_party/gemma/model.py [hunk 5]
```diff
@@ -35,6 +35,142 @@ def precompute_freqs_cis(
...
+  freqs_cis = reshape_for_broadcast(freqs_cis, x_)
+  x_out = torch.view_as_real(x_ * freqs_cis).type_as(x)
+  x_out = torch.cat(torch.chunk(x_out, 2, dim=-1), dim=-2)
+  x_out = x_out.reshape(
+      x_out.shape[0], x_out.shape[1], x_out.shape[2], -1
+  ).transpose(1, 2)
+  return x_out
+
+
+class GemmaAttention(nn.Module):
```
Comment (FanhaiLu1) [inline jetstream_pt/third_party/gemma/model.py:66]: In long term, we might need to extend from Attention class, large percentage of code are similar.

**Review by FanhaiLu1 — APPROVED** [review]

**Pull request merged**